A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+Admin Interface API
+----
+* Author(s): lidiz
+* Approver: markroth
+* Status: Draft
+* Implemented in: C++, Golang, Java
+* Last updated: 2021-01-27
+* Discussion at: TBD
+
+## Abstract
+
+This proposal describes a convenience API in each gRPC language to improve the usability of creating a gRPC server with admin services.
+
+
+## Background
+
+gRPC Admin Interface is part of the Proxyless Service Mesh Observability project which includes new gRPC admin services and a CLI tool (yet to release). The admin services are responsible for exposing states of the gRPC library via gRPC protocol, and the CLI tool will help users to better interpret the fetched data from the gRPC application. Currently, the existing admin server should at least expose two admin services:
+
+* Channelz: the channel tracing library that collects stats about a channel’s traffic, underlying sockets, and credentials ([proto definition](https://github.com/grpc/grpc/blob/master/src/proto/grpc/channelz/channelz.proto));
+* [Client Status Discovery Service](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/status/v3/csds.proto): the xDS config dump library that dumps the accepted xDS configuration with config status;
+
+However, the number of admin services might increase in the future. Solutions with no code changes, like env, or config files, are not desired (see rationale section below). On the other hand, as a library, gRPC would like to promote the usage of admin services. So, there is a need for an ideally one LOC to create and start an admin server.
+
+In addition, by providing new APIs, we abstract away the individual services with the concept admin services. We can also achieve **dynamic admin services selection**, for example, we **only add CSDS service if the xDS package is pulled in**. In this way, we won’t force users to take unnecessary dependencies.
+
+
+### Types of Services
+
+To help communicate the scenarios, we can define all services into several types, from the view of administrative functions:
+
+* **Primary service**: the user-provided services that handle the application logic;
```

Comment:
I don't think this taxonomy is useful in this document.  This document deals only with binary-global services, and we don't even need that term, since we don't have to mention the other two types, and we therefore don't need to differentiate this one type from the others.

Reply:
Removed. This was used to discuss whether we should recommend users to use separate port or the same serving port.